File: samples/csharp/src/AgentPatterns/Evaluation/EnsembleJudge.cs
using Microsoft.SemanticKernel.ChatCompletion;
namespace AgentPatterns.Evaluation;
public record EnsembleResult(string FinalVerdict, int Approvals, int Total);
public class EnsembleJudge {
    public static async Task<EnsembleResult> Evaluate(IList<IChatCompletionService> judges, string output, string task) {
        var verdicts = await Task.WhenAll(judges.Select(async judge => {
            var h = new ChatHistory($"Evaluate output for '{task}'. Output: {output}. Reply APPROVED or REJECTED.");
            var v = (await judge.GetChatMessageContentsAsync(h))[0].Content ?? "REJECTED";
            return v.ToUpper().Contains("APPROVED");
        }));
        var approvals = verdicts.Count(v => v);
        var verdict = approvals * 2 >= judges.Count ? "APPROVED" : "REJECTED";
        return new EnsembleResult(verdict, approvals, judges.Count);
    }
}

File: samples/java/src/main/java/com/agents/patterns/evaluation/EnsembleJudge.java
package com.agents.patterns.evaluation;

import org.springframework.ai.chat.client.ChatClient;

import java.util.List;

/**
 * Pattern: Ensemble Judge
 * Multiple independent judge LLMs each evaluate the same output.
 * The final verdict is decided by majority vote. This reduces single-judge bias
 * and increases confidence in the evaluation result.
 */
public class EnsembleJudge {

    public record EnsembleResult(String finalVerdict, int approvals, int total) {}

    /**
     * Collects verdicts from all judge LLMs and returns the majority decision.
     *
     * @param judges  list of judge ChatClients
     * @param output  the agent output to evaluate
     * @param task    the original task description (for context)
     * @return an EnsembleResult with the final verdict and vote counts
     */
    public static EnsembleResult evaluate(List<ChatClient> judges, String output, String task) {
        int approvals = 0;

        for (int i = 0; i < judges.size(); i++) {
            String judgePrompt = "You are Judge " + (i + 1) + ". Evaluate this output for task: "
                    + task + "\nOutput: " + output
                    + "\nReply with exactly one word: APPROVED or REJECTED.";
            String verdict = judges.get(i).prompt()
                    .user(judgePrompt)
                    .call()
                    .content()
                    .trim()
                    .toUpperCase();

            if (verdict.contains("APPROVED")) {
                approvals++;
            }
        }

        int total = judges.size();
        String finalVerdict = (approvals > total / 2) ? "APPROVED" : "REJECTED";
        return new EnsembleResult(finalVerdict, approvals, total);
    }
}
